homeassistant/components/proximity/coordinator.py

```diff
@@ -0,0 +1,268 @@
+    old_state: State | None
+    new_state: State | None
+
+
+class ProximityData(TypedDict):
+    """ProximityData type class."""
+
+    dist_to_zone: str | float
+    dir_of_travel: str | float
+    nearest: str | float
+
+
+class ProximityDataUpdateCoordinator(DataUpdateCoordinator[ProximityData]):
+    """Proximity data update coordinator."""
+
+    def __init__(
+        self, hass: HomeAssistant, friendly_name: str, config: ConfigType
+    ) -> None:
+        """Initialize the Proximity coordinator."""
+        self.ignored_zones: list[str] = config[CONF_IGNORED_ZONES]
+        self.proximity_devices: list[str] = config[CONF_DEVICES]
+        self.tolerance: int = config[CONF_TOLERANCE]
+        self.proximity_zone: str = config[CONF_ZONE]
+        self.unit_of_measurement: str = config.get(
+            CONF_UNIT_OF_MEASUREMENT, hass.config.units.length_unit
+        )
+        self.friendly_name = friendly_name
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=friendly_name,
+            update_interval=None,
+        )
+
+        self.data = {
+            "dist_to_zone": DEFAULT_DIST_TO_ZONE,
+            "dir_of_travel": DEFAULT_DIR_OF_TRAVEL,
+            "nearest": DEFAULT_NEAREST,
+        }
+
+        self.state_change_data: StateChangedData | None = None
+        async_track_state_change(
+            hass, self.proximity_devices, self.async_check_proximity_state_change
+        )
+
+    async def async_check_proximity_state_change(
+        self, entity: str, old_state: State | None, new_state: State | None
+    ) -> None:
+        """Fetch and process state change event."""
+        if new_state is None:
+            _LOGGER.debug("no new_state -> abort")
+            return
+
+        # We can't check proximity because latitude and longitude don't exist.
+        if "latitude" not in new_state.attributes:
+            _LOGGER.debug("no latitude and longitude -> abort")
+            return
+
+        self.state_change_data = StateChangedData(entity, old_state, new_state)
+        await self.async_refresh()
+
+    async def _async_update_data(self) -> ProximityData:
+        """Calculate Proximity data."""
+        if self.state_change_data is None or self.state_change_data.new_state is None:
+            return self.data
+
+        entity_name = self.state_change_data.new_state.name
+        devices_to_calculate = False
+        devices_in_zone = ""
+
+        zone_state = self.hass.states.get(f"zone.{self.proximity_zone}")
+        proximity_latitude = (
+            zone_state.attributes.get(ATTR_LATITUDE) if zone_state else None
+        )
+        proximity_longitude = (
+            zone_state.attributes.get(ATTR_LONGITUDE) if zone_state else None
+        )
+
+        # Check for devices in the monitored zone.
+        for device in self.proximity_devices:
+            if (device_state := self.hass.states.get(device)) is None:
+                devices_to_calculate = True
```

Comment:
this seems to be wrong and will be fixed in another PR which will also include additional tests for all these "abort condifitions"